File: android/app/src/main/kotlin/com/blindassist/blind_assist_app/channels/SensorChannel.kt
package com.blindassist.blind_assist_app.channels

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorManager
import android.util.Log
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * 传感器 Platform Channel
 * 
 * 主要提供 Flutter 插件无法直接获取的高级传感器能力，目前包括：
 * - 查询设备是否具备某类传感器
 * - 获取传感器详细参数（精度、分辨率等）
 * 
 * 实时传感器数据的读取由 sensors_plus / geolocator 等 Flutter 插件负责。
 * 
 * Channel 名称：com.blindassist/sensors
 */
class SensorChannel(
    private val context: Context,
    messenger: BinaryMessenger
) : MethodChannel.MethodCallHandler {

    companion object {
        private const val TAG = "SensorChannel"
        private const val CHANNEL_NAME = "com.blindassist/sensors"
    }

    private val channel = MethodChannel(messenger, CHANNEL_NAME)
    private val sensorManager: SensorManager =
        context.getSystemService(Context.SENSOR_SERVICE) as SensorManager

    init {
        channel.setMethodCallHandler(this)
        Log.i(TAG, "SensorChannel 已注册")
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "getAvailableSensors" -> {
                val sensors = listAvailableSensors()
                result.success(sensors)
            }
            "hasSensor" -> {
                val type = call.argument<String>("type") ?: ""
                result.success(hasSensor(type))
            }
            "getSensorInfo" -> {
                val type = call.argument<String>("type") ?: ""
                result.success(getSensorInfo(type))
            }
            else -> result.notImplemented()
        }
    }

    private fun listAvailableSensors(): List<Map<String, Any?>> {
        val sensors = sensorManager.getSensorList(Sensor.TYPE_ALL)
        return sensors.map { sensor ->
            mapOf(
                "name" to sensor.name,
                "type" to sensor.type,
                "typeName" to sensorTypeName(sensor.type),
                "vendor" to sensor.vendor,
                "maxRange" to sensor.maximumRange,
                "resolution" to sensor.resolution,
                "power" to sensor.power
            )
        }
    }

    private fun hasSensor(type: String): Boolean {
        val sensorType = mapTypeNameToInt(type) ?: return false
        return sensorManager.getDefaultSensor(sensorType) != null
    }

    private fun getSensorInfo(type: String): Map<String, Any?>? {
        val sensorType = mapTypeNameToInt(type) ?: return null
        val sensor = sensorManager.getDefaultSensor(sensorType) ?: return null
        return mapOf(
            "name" to sensor.name,
            "vendor" to sensor.vendor,
            "maxRange" to sensor.maximumRange,
            "resolution" to sensor.resolution,
            "power" to sensor.power,
            "minDelay" to sensor.minDelay
        )
    }

    private fun mapTypeNameToInt(type: String): Int? {
        return when (type.lowercase()) {
            "accelerometer" -> Sensor.TYPE_ACCELEROMETER
            "gyroscope" -> Sensor.TYPE_GYROSCOPE
            "magnetometer" -> Sensor.TYPE_MAGNETIC_FIELD
            "pressure" -> Sensor.TYPE_PRESSURE
            "light" -> Sensor.TYPE_LIGHT
            "proximity" -> Sensor.TYPE_PROXIMITY
            "step_counter" -> Sensor.TYPE_STEP_COUNTER
            "gravity" -> Sensor.TYPE_GRAVITY
            else -> null
        }
    }

    private fun sensorTypeName(type: Int): String {
        return when (type) {
            Sensor.TYPE_ACCELEROMETER -> "accelerometer"
            Sensor.TYPE_GYROSCOPE -> "gyroscope"
            Sensor.TYPE_MAGNETIC_FIELD -> "magnetometer"
            Sensor.TYPE_PRESSURE -> "pressure"
            Sensor.TYPE_LIGHT -> "light"
            Sensor.TYPE_PROXIMITY -> "proximity"
            Sensor.TYPE_STEP_COUNTER -> "step_counter"
            Sensor.TYPE_GRAVITY -> "gravity"
            else -> "unknown"
        }
    }

    fun dispose() {
        channel.setMethodCallHandler(null)
    }
}

File: ios/Runner/Channels/SensorChannel.swift
import Flutter
import CoreMotion
import UIKit

/// iOS 传感器 Platform Channel
///
/// 查询设备传感器硬件信息（由 sensors_plus 插件处理实时数据流）
///
/// Channel 名称：com.blindassist/sensors
class SensorChannel: NSObject {

    private static let channelName = "com.blindassist/sensors"
    private let channel: FlutterMethodChannel
    private let motionManager = CMMotionManager()

    init(messenger: FlutterBinaryMessenger) {
        self.channel = FlutterMethodChannel(name: Self.channelName, binaryMessenger: messenger)
        super.init()
        self.channel.setMethodCallHandler { [weak self] call, result in
            self?.handle(call: call, result: result)
        }
        NSLog("[SensorChannel] 已注册")
    }

    private func handle(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getAvailableSensors":
            result(availableSensors())

        case "hasSensor":
            let type = (call.arguments as? [String: Any])?["type"] as? String ?? ""
            result(hasSensor(type: type))

        case "getSensorInfo":
            let type = (call.arguments as? [String: Any])?["type"] as? String ?? ""
            result(getSensorInfo(type: type))

        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func availableSensors() -> [[String: Any?]] {
        var sensors: [[String: Any?]] = []
        if motionManager.isAccelerometerAvailable {
            sensors.append(sensorInfo(name: "Accelerometer", typeName: "accelerometer"))
        }
        if motionManager.isGyroAvailable {
            sensors.append(sensorInfo(name: "Gyroscope", typeName: "gyroscope"))
        }
        if motionManager.isMagnetometerAvailable {
            sensors.append(sensorInfo(name: "Magnetometer", typeName: "magnetometer"))
        }
        if motionManager.isDeviceMotionAvailable {
            sensors.append(sensorInfo(name: "DeviceMotion", typeName: "device_motion"))
        }
        // iOS 气压计
        if CMAltimeter.isRelativeAltitudeAvailable() {
            sensors.append(sensorInfo(name: "Barometer", typeName: "pressure"))
        }
        // iOS 步数计
        // 需要运动权限，此处只列出可用性
        return sensors
    }

    private func hasSensor(type: String) -> Bool {
        switch type.lowercased() {
        case "accelerometer": return motionManager.isAccelerometerAvailable
        case "gyroscope": return motionManager.isGyroAvailable
        case "magnetometer": return motionManager.isMagnetometerAvailable
        case "pressure": return CMAltimeter.isRelativeAltitudeAvailable()
        case "device_motion": return motionManager.isDeviceMotionAvailable
        default: return false
        }
    }

    private func getSensorInfo(type: String) -> [String: Any?]? {
        if hasSensor(type: type) {
            return sensorInfo(name: type, typeName: type)
        }
        return nil
    }

    private func sensorInfo(name: String, typeName: String) -> [String: Any?] {
        return [
            "name": name,
            "typeName": typeName,
            "vendor": "Apple",
            "maxRange": nil,
            "resolution": nil,
            "power": nil
        ]
    }

    deinit {
        channel.setMethodCallHandler(nil)
    }
}
